File: server/prosody.py
"""Per-turn prosody (pure numpy) for POST /analyze/upload.

Given the recording's PCM + each turn's start/end timestamps, we measure three
delivery signals and turn them into compact categorical labels:

* energy   — RMS amplitude over the turn window  → quiet / normal / loud
* pitch    — median F0 + F0 variability (frame-wise autocorrelation)
             → low / mid / high pitch, and flat / varied
* speech_rate — words / duration                 → slow / normal / fast

Two deliberate honesty rules:

1. Pitch is ``None`` when a turn is mostly unvoiced (silence, noise, non-speech):
   fewer than ``MIN_VOICED_FRACTION`` of frames show enough periodicity. We
   never invent an F0 for something that isn't voiced.

2. Labels are RELATIVE to *this recording's own* distribution (tertiles/median
   over its turns), not to any absolute dB/Hz scale. That is the correct choice
   here: microphone gain, distance, and device vary per recording, so "loud"
   can only mean "loud for this speaker on this recording". The LLM is told the
   labels are relative delivery cues, not absolute measurements.

Everything is a plain function over numpy arrays / dicts — trivially unit
testable with synthetic signals, no framework coupling.
"""

from __future__ import annotations

import numpy as np

# Frame-wise F0 estimation params. 40 ms frames / 10 ms hop is a standard
# pitch-analysis window; the 60–400 Hz search spans low male speech through
# high female/child speech.
FRAME_MS = 40.0
HOP_MS = 10.0
F0_MIN_HZ = 60.0
F0_MAX_HZ = 400.0
# A frame is "voiced" when its normalized autocorrelation peak in the F0 lag
# range clears this. A clean periodic tone sits near 1.0; broadband noise and
# silence sit well below.
VOICED_AUTOCORR_THRESHOLD = 0.35
# Below this fraction of voiced frames we report pitch as None (honest null).
MIN_VOICED_FRACTION = 0.20


# ---------------------------------------------------------------------------
# Single-turn feature extraction
# ---------------------------------------------------------------------------

def _turn_slice(pcm: np.ndarray, sr: int, start: float, end: float) -> np.ndarray:
    """The PCM samples spanning [start, end) seconds, clipped to the array."""
    if sr <= 0 or end <= start:
        return np.empty(0, dtype=np.float32)
    a = max(0, int(round(start * sr)))
    b = min(pcm.shape[0], int(round(end * sr)))
    if b <= a:
        return np.empty(0, dtype=np.float32)
    return pcm[a:b]


def rms_energy(samples: np.ndarray) -> float:
    """Root-mean-square amplitude (0.0 for an empty window)."""
    if samples.size == 0:
        return 0.0
    return float(np.sqrt(np.mean(np.square(samples, dtype=np.float64))))


def _frame_f0(frame: np.ndarray, sr: int) -> float | None:
    """Estimate one frame's F0 via autocorrelation, or ``None`` if unvoiced.

    The frame is mean-removed, autocorrelated, and normalized by its zero-lag
    energy. The strongest peak within the 60–400 Hz lag window decides voicing:
    a normalized peak below ``VOICED_AUTOCORR_THRESHOLD`` means "not periodic
    enough" → unvoiced → ``None``.
    """
    frame = frame.astype(np.float64)
    frame = frame - frame.mean()
    energy = float(np.dot(frame, frame))
    if energy <= 0.0:
        return None
    corr = np.correlate(frame, frame, mode="full")[frame.size - 1:]
    min_lag = int(sr / F0_MAX_HZ)
    max_lag = int(sr / F0_MIN_HZ)
    max_lag = min(max_lag, corr.size - 1)
    if max_lag <= min_lag:
        return None
    window = corr[min_lag:max_lag + 1]
    peak_offset = int(np.argmax(window))
    peak_lag = min_lag + peak_offset
    if peak_lag <= 0:
        return None
    if corr[peak_lag] / energy < VOICED_AUTOCORR_THRESHOLD:
        return None
    return sr / peak_lag


def estimate_pitch(
    samples: np.ndarray, sr: int,
) -> tuple[float | None, float | None, float]:
    """(median_f0, f0_std, voiced_fraction) over a turn's samples.

    ``median_f0`` and ``f0_std`` are ``None`` when fewer than
    ``MIN_VOICED_FRACTION`` of frames are voiced — an honest null for
    silence/noise rather than a fabricated pitch.
    """
    if sr <= 0 or samples.size == 0:
        return None, None, 0.0
    frame_len = max(1, int(sr * FRAME_MS / 1000.0))
    hop = max(1, int(sr * HOP_MS / 1000.0))
    if samples.size < frame_len:
        return None, None, 0.0

    f0s: list[float] = []
    n_frames = 0
    for start in range(0, samples.size - frame_len + 1, hop):
        n_frames += 1
        f0 = _frame_f0(samples[start:start + frame_len], sr)
        if f0 is not None:
            f0s.append(f0)

    if n_frames == 0:
        return None, None, 0.0
    voiced_fraction = len(f0s) / n_frames
    if voiced_fraction < MIN_VOICED_FRACTION or not f0s:
        return None, None, voiced_fraction
    arr = np.asarray(f0s, dtype=np.float64)
    return float(np.median(arr)), float(np.std(arr)), voiced_fraction


def turn_features(pcm: np.ndarray, sr: int, start: float, end: float) -> dict:
    """Raw prosody numbers for one turn window (no labels yet).

    Keys: ``rms``, ``f0_median`` (None if unvoiced), ``f0_std`` (None if
    unvoiced), ``voiced_fraction``.
    """
    samples = _turn_slice(pcm, sr, start, end)
    f0_median, f0_std, voiced_fraction = estimate_pitch(samples, sr)
    return {
        "rms": rms_energy(samples),
        "f0_median": f0_median,
        "f0_std": f0_std,
        "voiced_fraction": voiced_fraction,
    }


# ---------------------------------------------------------------------------
# Relative labeling across the recording's own turns
# ---------------------------------------------------------------------------


# The exact label vocabularies the mobile client types against
# (apps/mobile/src/api/client.ts `Voice`). Locked by test — do not drift.
ENERGY_LABELS = ("quiet", "normal", "loud")
PITCH_LABELS = ("low", "mid", "high")
RATE_LABELS = ("slow", "normal", "fast")

def _tertile_label(
    value: float, values: list[float], labels: tuple[str, str, str],
) -> str:
    """Label ``value`` low/mid/high by its position in ``values``' tertiles.

    Uses the 33rd/66th percentiles of the recording's own distribution — so
    labels are always relative to this recording. Degenerate distributions (all
    equal) collapse to the middle label.
    """
    arr = np.asarray(values, dtype=np.float64)
    lo, hi = np.percentile(arr, [100.0 / 3.0, 200.0 / 3.0])
    if lo == hi:
        # Degenerate distribution (all values equal, or a single turn): nothing
        # stands out, so everything is baseline — the middle label. Without
        # this, `value <= lo` labels every turn of an evenly-delivered
        # recording "quiet"/"slow", which the UI then renders as noteworthy.
        return labels[1]
    if value <= lo:
        return labels[0]
    if value <= hi:
        return labels[1]
    return labels[2]


def label_turns(
    features: list[dict], transcript_turns: list[dict],
) -> list[dict]:
    """Compact per-turn voice labels + the raw numbers behind them.

    ``features[i]`` is a :func:`turn_features` dict; ``transcript_turns[i]``
    supplies the text (for speech rate) and timestamps (for duration). Each
    output dict carries:

    * ``energy_label``  — quiet / normal / loud   (relative to this recording)
    * ``pitch_label``   — low / mid / high pitch, or ``None`` when unvoiced
    * ``pitch_var_label`` — flat / varied, or ``None`` when unvoiced
    * ``rate_label``    — slow / normal / fast
    * raw ``rms``, ``f0_median``, ``f0_std``, ``speech_rate``

    Energy/pitch/rate are labelled by tertiles over the recording's own turns
    (see module docstring for why relative is correct here).
    """
    n = len(features)
    rms_values = [f["rms"] for f in features]

    # Speech rate = words / duration; duration from the turn's own timestamps.
    rates: list[float] = []
    for f, t in zip(features, transcript_turns):
        start = t.get("start_time") or 0.0
        end = t.get("end_time") or 0.0
        duration = end - start
        words = len((t.get("text") or "").split())
        rates.append(words / duration if duration > 0 else 0.0)

    # Pitch level + variability are only meaningful over VOICED turns; unvoiced
    # turns get a None label rather than being forced onto the scale.
    voiced_idx = [i for i in range(n) if features[i]["f0_median"] is not None]
    voiced_f0 = [features[i]["f0_median"] for i in voiced_idx]
    voiced_std = [features[i]["f0_std"] for i in voiced_idx]
    std_median = float(np.median(voiced_std)) if voiced_std else 0.0

    out: list[dict] = []
    for i in range(n):
        f = features[i]
        energy_label = _tertile_label(
            f["rms"], rms_values, ENERGY_LABELS,
        )
        rate_label = _tertile_label(
            rates[i], rates, RATE_LABELS,
        )
        if f["f0_median"] is None:
            pitch_label = None
            pitch_var_label = None
        else:
            pitch_label = _tertile_label(
                f["f0_median"], voiced_f0,
                PITCH_LABELS,
            )
            # Binary flat/varied around the recording's median F0 spread.
            pitch_var_label = "varied" if f["f0_std"] > std_median else "flat"
        out.append({
            "energy_label": energy_label,
            "pitch_label": pitch_label,
            "pitch_var_label": pitch_var_label,
            "rate_label": rate_label,
            "rms": round(f["rms"], 6),
            "f0_median": (
                None if f["f0_median"] is None else round(f["f0_median"], 2)
            ),
            "f0_std": None if f["f0_std"] is None else round(f["f0_std"], 2),
            "speech_rate": round(rates[i], 3),
        })
    return out


def annotate(labels: dict) -> str:
    """One-line delivery annotation for the LLM prompt, e.g. ``loud, fast,
    pitch varied``. Built purely from the labels — no second model call."""
    parts = [labels["energy_label"], labels["rate_label"]]
    if labels["pitch_var_label"] is not None:
        parts.append(f"pitch {labels['pitch_var_label']}")
    else:
        parts.append("pitch unclear")
    return ", ".join(parts)

File: apps/mobile/src/live/prosody.ts
/**
 * Per-turn prosody, on the phone — a straight port of server/prosody.py's
 * feature extraction (RMS energy, autocorrelation pitch, speech-rate proxy).
 *
 * Only the MEASUREMENT half is ported. The server's relative labelling
 * (tertiles over a whole recording's turns) needs the whole recording; the
 * realtime path reports raw numbers in `TurnLocalEvent.prosody` and lets the
 * server/episode record label them later. Same honesty rule as the server:
 * pitch is `null` when a turn is mostly unvoiced — never a fabricated F0.
 *
 * Pure functions over Float32Array — no native seam, fully Jest-testable on
 * synthetic tones (see __tests__/liveProsody.test.ts, which mirrors
 * server/tests/test_prosody.py's expectations).
 */

// Frame-wise F0 params — identical to server/prosody.py so both runtimes
// measure the same thing.
export const FRAME_MS = 40.0;
export const HOP_MS = 10.0;
export const F0_MIN_HZ = 60.0;
export const F0_MAX_HZ = 400.0;
export const VOICED_AUTOCORR_THRESHOLD = 0.35;
export const MIN_VOICED_FRACTION = 0.2;

/** Root-mean-square amplitude (0 for an empty window). */
export function rmsEnergy(samples: Float32Array): number {
  if (samples.length === 0) return 0;
  let acc = 0;
  for (let i = 0; i < samples.length; i++) acc += samples[i] * samples[i];
  return Math.sqrt(acc / samples.length);
}

/** RMS loudness in dB relative to full scale for float PCM in [-1, 1];
 *  `-Infinity` for digital silence (matches watch.vectors.rms_dbfs, which
 *  divides int16 by 32768 — same number for the same signal). */
export function rmsDbfs(samples: Float32Array): number {
  const rms = rmsEnergy(samples);
  if (rms <= 0) return -Infinity;
  return 20 * Math.log10(rms);
}

/**
 * One frame's F0 via normalized autocorrelation, or `null` when unvoiced.
 * Mean-removed, autocorrelated over the 60–400 Hz lag window, normalized by
 * the zero-lag energy; a peak below VOICED_AUTOCORR_THRESHOLD is "not
 * periodic enough". Same decision rule as server/prosody.py::_frame_f0 —
 * only the lags actually needed are computed (the numpy version correlates
 * over every lag and then slices).
 */
export function frameF0(frame: Float32Array, sr: number): number | null {
  const n = frame.length;
  if (n === 0) return null;
  let mean = 0;
  for (let i = 0; i < n; i++) mean += frame[i];
  mean /= n;
  const x = new Float64Array(n);
  let energy = 0;
  for (let i = 0; i < n; i++) {
    x[i] = frame[i] - mean;
    energy += x[i] * x[i];
  }
  if (energy <= 0) return null;
  const minLag = Math.floor(sr / F0_MAX_HZ);
  let maxLag = Math.floor(sr / F0_MIN_HZ);
  maxLag = Math.min(maxLag, n - 1);
  if (maxLag <= minLag) return null;
  let peakLag = -1;
  let peak = -Infinity;
  for (let lag = minLag; lag <= maxLag; lag++) {
    let c = 0;
    for (let i = 0; i + lag < n; i++) c += x[i] * x[i + lag];
    // Strict > keeps the FIRST maximum, matching numpy argmax on ties.
    if (c > peak) {
      peak = c;
      peakLag = lag;
    }
  }
  if (peakLag <= 0) return null;
  if (peak / energy < VOICED_AUTOCORR_THRESHOLD) return null;
  return sr / peakLag;
}

export interface PitchEstimate {
  /** Median F0 over voiced frames; null when < MIN_VOICED_FRACTION voiced. */
  f0Median: number | null;
  /** Population std-dev of the voiced F0s (numpy's default ddof=0); null with f0Median. */
  f0Std: number | null;
  voicedFraction: number;
}

function median(values: number[]): number {
  const s = [...values].sort((a, b) => a - b);
  const mid = s.length >> 1;
  return s.length % 2 ? s[mid] : (s[mid - 1] + s[mid]) / 2;
}

function pitchFromF0s(f0s: number[], nFrames: number): PitchEstimate {
  if (nFrames === 0) return { f0Median: null, f0Std: null, voicedFraction: 0 };
  const voicedFraction = f0s.length / nFrames;
  if (voicedFraction < MIN_VOICED_FRACTION || f0s.length === 0) {
    return { f0Median: null, f0Std: null, voicedFraction };
  }
  const mean = f0s.reduce((a, b) => a + b, 0) / f0s.length;
  const variance = f0s.reduce((a, b) => a + (b - mean) * (b - mean), 0) / f0s.length;
  return { f0Median: median(f0s), f0Std: Math.sqrt(variance), voicedFraction };
}

/** Port of server/prosody.py::estimate_pitch. */
export function estimatePitch(samples: Float32Array, sr: number): PitchEstimate {
  if (sr <= 0 || samples.length === 0) {
    return { f0Median: null, f0Std: null, voicedFraction: 0 };
  }
  const frameLen = Math.max(1, Math.floor((sr * FRAME_MS) / 1000));
  const hop = Math.max(1, Math.floor((sr * HOP_MS) / 1000));
  if (samples.length < frameLen) {
    return { f0Median: null, f0Std: null, voicedFraction: 0 };
  }
  const f0s: number[] = [];
  let nFrames = 0;
  for (let start = 0; start + frameLen <= samples.length; start += hop) {
    nFrames += 1;
    const f0 = frameF0(samples.subarray(start, start + frameLen), sr);
    if (f0 !== null) f0s.push(f0);
  }
  return pitchFromF0s(f0s, nFrames);
}

/** Frames of pitch analysis between yields to the event loop in
 *  `estimatePitchAsync` (~0.5 s of audio; a few ms of work on a JIT, tens
 *  on Hermes). */
export const PITCH_YIELD_EVERY_FRAMES = 50;

export interface AsyncPitchOptions {
  /** Yield after this many frames (default PITCH_YIELD_EVERY_FRAMES). */
  yieldEvery?: number;
  /** How to yield (default: a 0 ms timer, i.e. a macrotask boundary). */
  sleep?: (ms: number) => Promise<void>;
}

const defaultSleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

/**
 * `estimatePitch`, numerically identical, but cooperative: the per-frame
 * autocorrelation is the one O(n·lags) loop in the live path, and on the
 * phone it shares the JS thread with the mic callbacks, the WebSocket and
 * the UI. Yielding every few dozen frames keeps audio flowing while a long
 * turn is measured.
 */
export async function estimatePitchAsync(
  samples: Float32Array,
  sr: number,
  opts: AsyncPitchOptions = {},
): Promise<PitchEstimate> {
  if (sr <= 0 || samples.length === 0) {
    return { f0Median: null, f0Std: null, voicedFraction: 0 };
  }
  const frameLen = Math.max(1, Math.floor((sr * FRAME_MS) / 1000));
  const hop = Math.max(1, Math.floor((sr * HOP_MS) / 1000));
  if (samples.length < frameLen) {
    return { f0Median: null, f0Std: null, voicedFraction: 0 };
  }
  const yieldEvery = Math.max(1, opts.yieldEvery ?? PITCH_YIELD_EVERY_FRAMES);
  const sleep = opts.sleep ?? defaultSleep;
  const f0s: number[] = [];
  let nFrames = 0;
  for (let start = 0; start + frameLen <= samples.length; start += hop) {
    nFrames += 1;
    const f0 = frameF0(samples.subarray(start, start + frameLen), sr);
    if (f0 !== null) f0s.push(f0);
    if (nFrames % yieldEvery === 0) await sleep(0);
  }
  return pitchFromF0s(f0s, nFrames);
}

/** The numbers `TurnLocalEvent.prosody` carries (server/models/audio.py
 *  TurnProsody). Every field nullable: on-device measurement is best-effort. */
export interface TurnProsody {
  rms_dbfs: number | null;
  pitch_hz: number | null;
  speech_rate: number | null;
}

/**
 * Measure one finalized turn. `text` (from on-device STT) and the turn's
 * duration give the speech-rate proxy (words / second — the same words-over-
 * duration definition server/prosody.py::label_turns uses); null when there is
 * no text or no duration to divide by, never 0-as-a-guess.
 */
export function turnProsody(
  samples: Float32Array,
  sr: number,
  text: string,
  durationSeconds: number,
): TurnProsody {
  const dbfs = rmsDbfs(samples);
  const { f0Median } = estimatePitch(samples, sr);
  return assembleProsody(dbfs, f0Median, text, durationSeconds);
}

function assembleProsody(
  dbfs: number,
  f0Median: number | null,
  text: string,
  durationSeconds: number,
): TurnProsody {
  const words = text.trim() === "" ? 0 : text.trim().split(/\s+/).length;
  const speechRate =
    words > 0 && durationSeconds > 0 ? words / durationSeconds : null;
  return {
    rms_dbfs: Number.isFinite(dbfs) ? Math.round(dbfs * 100) / 100 : null,
    pitch_hz: f0Median === null ? null : Math.round(f0Median * 100) / 100,
    speech_rate: speechRate === null ? null : Math.round(speechRate * 1000) / 1000,
  };
}

/** Pitch is measured over at most the LAST this-many seconds of a turn in
 *  the live path: the per-frame autocorrelation costs ~100k multiply-adds a
 *  frame (100 frames/s), so an unbounded monologue would stall the JS
 *  thread for seconds; the median F0 of the last few seconds is the same
 *  delivery signal. Loudness is still over the whole turn (O(n)). */
export const LIVE_MAX_PITCH_SECONDS = 6;

export interface AsyncProsodyOptions extends AsyncPitchOptions {
  /** Cap on the pitch window (seconds from the END of the turn);
   *  undefined/Infinity = whole turn. */
  maxPitchSeconds?: number;
}

/** `turnProsody` for the live loop: same numbers, cooperative and bounded. */
export async function turnProsodyAsync(
  samples: Float32Array,
  sr: number,
  text: string,
  durationSeconds: number,
  opts: AsyncProsodyOptions = {},
): Promise<TurnProsody> {
  const dbfs = rmsDbfs(samples);
  const cap = opts.maxPitchSeconds;
  const window =
    cap !== undefined && Number.isFinite(cap) && cap > 0 && samples.length > cap * sr
      ? samples.subarray(samples.length - Math.round(cap * sr))
      : samples;
  const { f0Median } = await estimatePitchAsync(window, sr, opts);
  return assembleProsody(dbfs, f0Median, text, durationSeconds);
}
